### PR TITLE
[7.0] Split configuration & routes can be cached from Application contract.

### PR DESCRIPTION
**Pull request by crynobone**
Follow-up to #29789 based on laravel/ideas#1772 

Add check under `Illuminate\Support\ServiceProvider` to check against specific contracts. With this we can remove the methods in Lumen since it doesn't support `CachesConfiguration` or `CachesRoutes`.

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

### CODE REVIEW

File: src/Illuminate/Broadcasting/BroadcastManager.php [hunk 1]
```diff
@@ -21,7 +22,7 @@ class BroadcastManager implements FactoryContract
     /**
      * The application instance.
      *
-     * @var \Illuminate\Contracts\Foundation\Application
+     * @var \Illuminate\Contracts\Container\Container
      */
     protected $app;
```
Comment (crynobone) [inline src/Illuminate/Broadcasting/BroadcastManager.php]: Other than checking for `routesAreCached()`, the only used for `$app` here are related to `Container` contract to resolve `request`, `events`, `queue`, `redis`, `config` and `Psr\Log\LoggerInterface`.

**Review by driesvints — APPROVED** [review]
Yes! 👍

**Comment by GrahamCampbell** [comment]
Super nice change. 🚀

**Comment by taylorotwell** [comment]
Just so I understand. What is the goal here? What are we trying to achieve with all of these application contract changes?

**Comment by driesvints** [comment]
@taylorotwell the idea is that when people are trying to use illuminate components that require functionality normally provided by the Application class they can inject smaller implementations other than having to implement the entire Application contract.

We're grouping functionality under smaller contracts so they're both easier to implement and are more easier to digest than one big gigantic contract. In my ([and a lot of other's](https://github.com/laravel/ideas/issues/1772)) eyes it doesn't makes sense to have one big Application contract because that basically just says "swap Laravel with a different framework core". Check the ideas issue linked above by @crynobone for more ideas and info.

**Pull request merged**